### PR TITLE
Add type hints to fluent.syntax & fluent.runtime

### LINKED ISSUE #165: Add type hinting
Currently, mypy checking of my python code fails with:

```
 error: Skipping analyzing 'fluent.runtime': found module but no type hints or library stubs
```

I'll just ignore the error for now but it would be nice to have type hint support. It looks like this is blocked by https://github.com/projectfluent/python-fluent/issues/162.

### PR DESCRIPTION
**Pull request by eemeli**
Fixes #165 

The work here is split into commits that may be easier to review separately. The internal hints for the runtime resolver and compiler are pretty weak, as the base implementation is making explicit use of the language being fundamentally untyped.

The bugs that are fixed are all various sorts of corner cases that the code didn't account for previously; the errors tend to follow similar cases in fluent.js.

While adding the types I used both mypy and pylance; the added CI tests only apply `mypy --strict` as the strict-mode pylance checks don't all pass (mostly due to using `Unknown` values in places that accept `Any`).

### CODE REVIEW

**Comment by mitsuaky** [comment]
Are you planning on bumping to 0.19 after this?

**Comment by eemeli** [comment]
Yes.

**Review by eviljeff — COMMENTED** [review]
Just random nits/comments that should be taken as suggestions at most.

File: fluent.runtime/fluent/runtime/types.py [hunk 1]
```diff
@@ -65,30 +68,34 @@ class NumberFormatOptions:
...
-    currency = attr.ib(default=None)
-    currencyDisplay = attr.ib(default=CURRENCY_DISPLAY_SYMBOL,
-                              validator=attr.validators.in_(CURRENCY_DISPLAY_OPTIONS))
-    useGrouping = attr.ib(default=True)
-    minimumIntegerDigits = attr.ib(default=None)
-    minimumFractionDigits = attr.ib(default=None)
-    maximumFractionDigits = attr.ib(default=None)
-    minimumSignificantDigits = attr.ib(default=None)
-    maximumSignificantDigits = attr.ib(default=None)
+    style: Literal['decimal', 'currency', 'percent'] = attr.ib(
```
[the remark this answers is not in the document]
Comment (eemeli) [reply fluent.runtime/fluent/runtime/types.py:71]: I don't think that's possible with `Literal[]`, or at least I'm not aware of a way of an alternative type expression that would resolve to this.

File: fluent.runtime/fluent/runtime/resolver.py [hunk 2]
```diff
@@ -256,44 +300,55 @@ def select_from_select_expression(self, env, key):
...
     if is_number(val1):
         if not is_number(val2):
             # Could be plural rule match
-            return env.context._plural_form(val1) == val2
+            return cast(bool, env.context._plural_form(val1) == val2)
     elif is_number(val2):
         return match(val2, val1, env)
 
-    return val1 == val2
+    return cast(bool, val1 == val2)
```
[the remark this answers is not in the document]
Comment (eemeli) [reply fluent.runtime/fluent/runtime/resolver.py:326]: It's because the inputs are `Any`, and that at least technically allows for an `__eq__(self, other)` overload that returns something other than a `bool`.
